File: day12/src/p2.kt
import java.io.BufferedReader
import java.io.File

enum class Direction {
    Top, Bottom, Left, Right
}

class Position(val x: Int, val y: Int) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Position) return false
        return this.x == other.x && this.y == other.y
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}

class PerimeterData(
    val position: Position,
    val exitTop: Boolean,
    val exitRight: Boolean,
    val exitBottom: Boolean,
    val exitLeft: Boolean
) {}

fun part2() {
    val bufferedReader: BufferedReader = File("input.test.txt").bufferedReader()
    val inputString = bufferedReader.use { it.readText() }
    val matrix = inputString
        .replace("\r", "")
        .split("\n")

    val h = matrix.size
    val w = matrix[0].length

    val perimeters = mutableMapOf<Int, MutableList<PerimeterData>>()
    val areas = mutableMapOf<Int, Int>()
    val visited = mutableListOf<Pair<Int, Int>>()

    var currentID = 0

    fun posIsEdge(position: Position): Boolean {
        return position.x == 0 || position.y == 0 || position.x == w - 1 || position.y == h - 1
    }

    fun posExists(position: Position): Boolean {
        return position.x >= 0 && position.y >= 0 && position.x < w && position.y < h
    }

    fun checkPosition(pos: Position) {
        val x = pos.x
        val y = pos.y
        if (visited.contains(Pair(x, y))) {
            return
        }
        visited.add(Pair(x, y))
        areas[currentID] = areas[currentID]!! + 1

        // Check if we are on the edge of the matrix, if we are, we are perimeter
        if (posIsEdge(pos)) {
            perimeters[currentID]!!.add(
                PerimeterData(
                    position = pos,
                    exitTop = (y == 0),
                    exitRight = (x == w - 1),
                    exitBottom = (y == h - 1),
                    exitLeft = (x == 0)
                )
            )
        }

        for ((current, dir) in arrayOf(
            Pair(Position(x - 1, y), Direction.Left),
            Pair(Position(x + 1, y), Direction.Right),
            Pair(Position(x, y - 1), Direction.Top),
            Pair(Position(x, y + 1), Direction.Bottom),
        )) {
            if (posExists(current)) {
                if (matrix[current.y][current.x] == matrix[pos.y][pos.x]) {
                    checkPosition(current)
                } else {
                    perimeters[currentID]!!.add(
                        PerimeterData(
                            position = pos,
                            exitTop = dir == Direction.Top,
                            exitRight = dir == Direction.Right,
                            exitBottom = dir == Direction.Bottom,
                            exitLeft = dir == Direction.Left
                        )
                    )
                }
            }
        }
    }

    for (y in matrix.indices) {
        for (x in matrix[y].indices) {
            areas[currentID] = 0
            perimeters[currentID] = mutableListOf()
            checkPosition(Position(x, y))

            currentID++
        }
    }

    var sum = 0

    for (id in 0..currentID) {
        val area = areas[id]!!
        val perimeter = perimeters[id]!!
        val initialPos = perimeter.first { it.exitTop };
        var currentPos = initialPos
        var sideCount = 1
        var currentSide = Direction.Top
        do {
            val nextPos = when (currentSide) {
                Direction.Top -> Position(currentPos.position.x + 1, currentPos.position.y)
                Direction.Right -> Position(currentPos.position.x, currentPos.position.y + 1)
                Direction.Left -> Position(currentPos.position.x, currentPos.position.y - 1)
                Direction.Bottom -> Position(currentPos.position.x - 1, currentPos.position.y)
            }
            // if perimeter includes nextPos, we are still along the perimeter, if not we must turn and add the number of sides
            val nextPosIndex = perimeter.indexOfFirst { it.position == nextPos }
            if (nextPosIndex != -1) {
                currentPos = perimeter[nextPosIndex]
            } else {
                sideCount++
                currentSide = when (currentSide) {
                    Direction.Top -> Direction.Right
                    Direction.Right -> Direction.Bottom
                    Direction.Bottom -> Direction.Left
                    Direction.Left -> Direction.Top
                }
            }
        } while (!(currentPos.position.equals(initialPos.position)) && currentSide != Direction.Top)

        println("$sum + ($sideCount * $area) = ${sum + sideCount * area}")
        sum += sideCount * area
    }

    println("====================================")
    println("FINAL RESULT: $sum")
    println("====================================")
}

File: day12/p2.ts
import { readFile } from 'fs/promises';

type Position = {
  x: number, y: number
}

const posEq = (a: Position, b: Position) => {
  return a.x === b.x && a.y === b.y;
}

(async () => {
  const input = await readFile('./input.txt', { encoding: 'utf8' });
  const matrix = input
    .replace('\r', '')
    .split('\n');

  const h = matrix.length;
  const w = matrix[0].length;

  const areas: Record<number, number> = {}
  const corners: Record<number, number> = {}
  const visited: Position[] = []

  let currentID = 0;

  const positionExists = (position: Position) => {
    return position.x >= 0 && position.y >= 0 && position.x < w && position.y < h
  }

  const valueOrNull = (p: Position) => {
    if (positionExists(p)) return matrix[p.y][p.x];
    return null;
  }

  const getNeighbours = (p: Position) => {
    return {
      tl: valueOrNull({ x: p.x - 1, y: p.y - 1 }),
      tm: valueOrNull({ x: p.x, y: p.y - 1 }),
      tr: valueOrNull({ x: p.x + 1, y: p.y - 1 }),
      ml: valueOrNull({ x: p.x - 1, y: p.y }),
      mr: valueOrNull({ x: p.x + 1, y: p.y }),
      bl: valueOrNull({ x: p.x - 1, y: p.y + 1 }),
      bm: valueOrNull({ x: p.x, y: p.y + 1 }),
      br: valueOrNull({ x: p.x + 1, y: p.y + 1 }),
    }
  }

  const checkPosition = (pos: Position, id: number) => {
    if (!positionExists(pos)) return;
    if (visited.findIndex((p) => posEq(p, pos)) !== -1) { return; }
    visited.push(pos);

    const value = valueOrNull(pos);
    areas[id] += 1
    const neighbourLocations = getNeighbours(pos);
    const neighbours = Object.fromEntries(
      Object.entries(neighbourLocations).map(([k, v]) => [k, v === value])
    );

    let cornerCount = 0;
    // TL
    if ((!neighbours.tl && neighbours.tm && neighbours.ml) || (!neighbours.ml && !neighbours.tm)) cornerCount++;
    // TR
    if ((!neighbours.tr && neighbours.tm && neighbours.mr) || (!neighbours.mr && !neighbours.tm)) cornerCount++;
    // BL
    if ((!neighbours.bl && neighbours.bm && neighbours.ml) || (!neighbours.ml && !neighbours.bm)) cornerCount++;
    // BR
    if ((!neighbours.br && neighbours.bm && neighbours.mr) || (!neighbours.mr && !neighbours.bm)) cornerCount++;

    if (cornerCount > 0) console.log({ cornerCount, pos, id });

    corners[id] += cornerCount;

    const { x, y } = pos;
    for (let dPos of [
      { x: x + 1, y: y },
      { x: x - 1, y: y },
      { x: x, y: y + 1 },
      { x: x, y: y - 1 },
    ]) {
      if (valueOrNull(dPos) === value) checkPosition(dPos, id);
    }
  }

  for (let y = 0; y < matrix.length; y++) {
    for (let x = 0; x < matrix[0].length; x++) {
      if (visited.findIndex((p) => posEq(p, { x, y })) !== -1) { continue; }
      areas[currentID] = 0;
      corners[currentID] = 0;
      checkPosition({ x, y }, currentID);

      currentID++;
    }
  }
  let sum = 0

  for (let id = 0; id < currentID; id++) {
    console.log({
      area: areas[id],
      corner: corners[id],
      sum
    });
    sum += areas[id] * corners[id];
  }

  console.log('==================================')
  console.log(sum);
  console.log('==================================')
})();
